### DISCUSSION TITLE
Guidance needed to improve rtsp streaming latency

### DISCUSSION THREAD

**Discussion by fuatakgun** [bluenviron/mediamtx#584] [Questions] [answered]
First of all, thank your for your super work, I have built a docker add on around your docker image to use with Home Assistant Open Source Automation Software, more specifically to enable streaming from incoming video chunks. https://github.com/fuatakgun/eufy_security#62-installing-rtsp-simple-server-add-on-for-faster-p2p-streaming

Data is coming to a central software as bytes and I am using ffmpeg to convert incoming data into your RTSP server, without any transcoding. But from the first message sent to RTSP server to start streaming, it requires 4-5 seconds.

My question;
- is there any way to reduce the initial latency from your docker perspective?
- from RTSP protocol perspective, can I initiate a topic without actually sending any message and keep it open? so, when messages are starting to arrive, I can continue writing on it, rather than opening the topic when it starts receiving messages. I believe, RTSP protocol requires some basic information about codec, frame rate etc, can I manually supply them at startup time?
- maybe, I can publish a black image all the time while video chunks are not coming, is this feasible from rtsp perspective?

Thanks again for your super useful work.

BACKGROUND: before using RTSP based streaming, I had tried file based streaming, in which, ffmpeg was writing the video output into file system and I was streaming over there. It was 3-4 seconds slower and it was eating up SD CARD life time on raspberrypi. So, eventually moved to your docker, which improved stability and latency.

**Comment by aler9** [comment]
Hello,

> is there any way to reduce the initial latency from your docker perspective?

Initial latency depends on the period between the H264 I-frames or the original stream. FFmpeg usually waits for the first I-frame before starting streaming. The server doesn't introduce any kind of latency or delay. Therefore, if you want to improve the latency, try to edit the camera settings and decrease the "key frame interval".

> can I initiate a topic without actually sending any message and keep it open? RTSP protocol requires some basic information about codec, frame rate etc, can I manually supply them at startup time?

Yes you can initiate a topic and keep it open through keepalives, but you have to provide track informations (basically the codec and its parameters) during the initial handshake. If i were you, i'd write a custom RTSP client. gortsplib allows you to do all of that, just start from this example:

https://github.com/aler9/gortsplib/blob/main/examples/client-publish/main.go

> I can publish a black image all the time while video chunks are not coming, is this feasible from rtsp perspective?

The question is whether you want to re-encode or not. If you don't want to re-encode and need to save CPU, then you can't insert any frame into the H264 stream that is already coming from the camera. If you're ok with re-encoding, then you can do it.

Anyway, black images are not needed to keep the RTSP connection open, since gortsplib (and most of the other RTSP implementations) already sends keepalive packets.

**Comment by fuatakgun** [comment]
thanks for your reply. one more follow up question, when a client asks for describe about a track into rtsp server, can we put timeout or wait period? In my case, describe and stream starting is happening next to each other, and sometimes, i am receiving `no one is publishing` even though it is just milliseconds away :)

**Comment by aler9** [comment, marked as answer]
> when a client asks for describe about a track into rtsp server, can we put timeout or wait period?

This is already implemented when `runOnDemand` is used, like this:

```yml
paths:
  mypath:
    runOnDemand: ffmpeg -re -i myfile.mp4 -c copy -f rtsp rtsp://localhost:8554/mypath
```

In this case, when a user requests mypath with a DESCRIBE request, there's a wait period that allows ffmpeg to connect.

Therefore, in your case, you can add a dummy `runOnDemand` command to activate this behavior; in this way, readers will wait for the publisher to connect before receiving a DESCRIBE response:

```yml
paths:
  mypath:
    runOnDemand: echo "i'm waiting for a publisher that will connect normally, not through runOnDemand"
```

[a comment by an automated account is not shown]
